Fix  recursive spinlock in k_msgq_get

lucasdietrich opened the pull request:
Proposition for [solution 2](https://github.com/zephyrproject-rtos/zephyr/issues/45267#issuecomment-1116361810) to fix issue #45267 + test

Fixes #45267

lucasdietrich committed:
kernel: workq: Add internal function z_work_submit_to_queue()

This adds the internal function z_work_submit_to_queue(), which
submits the work item to the queue but doesn't force the thread to yield,
compared to the public function k_work_submit_to_queue().

When called from poll.c in the context of k_work_poll events, it ensures
that the thread does not yield in the context of the spinlock of object
that became available.

Fixes #45267

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

andyross commented on kernel/work.c:
```diff
@@ -355,19 +355,38 @@ static int submit_to_queue_locked(struct k_work *work,
+	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_work, submit_to_queue, queue, work);
+
+	int ret = z_work_submit_to_queue(queue, work);
+
```
FWIW: I think this patch crossed with one of mine that just went in hitting exactly these lines (it replaces the yield with a proper reschedule), see commit https://github.com/zephyrproject-rtos/zephyr/commit/8d94967ec4773d9af67cb70167fe765085f3f737  Presumably it needs a rebase?

andyross commented on tests/kernel/workq/work_queue/src/main.c:
```diff
@@ -701,12 +725,79 @@ static void test_triggered_wait_expired(void)
+	zassert_true(k_msgq_get(&triggered_from_msgq_test.msgq, &msg, K_NO_WAIT) == 0, NULL);
+}
+
+static void test_triggered_from_msgq_init(void)
```
Pedantry, but this is one of those situations where you're testing for a regression that is going to seem really obscure to someone reading this code in the future.  Might be a good idea to explain it here, or just link to the github issue.

andyross reviewed (dismissed):
This definitely looks correct to me.  Giving it a +1 here for posterity, even though I'm sure you'll have to wipe that out with a rebase and I'll need to be bugged to come look at it again.

lucasdietrich committed:
tests: workq: Add a regression test for issue #45267

When an object availability event triggers a k_work_poll
item, the object lock should not be held anymore
during the execution of the work callback.

Signed-off-by: Lucas Dietrich <ld.adecy@gmail.com>

andyross reviewed (dismissed):
This looks good to me.  Note that:

> This solves https://github.com/zephyrproject-rtos/zephyr/issues/45267

If you rephrase that to "Fixes #45267" then the merge will automatically close the bug.

andyross reviewed (approved):
Refresh +1, looks good to me.
